scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+	 * priority (without considering boosting). Note that the ideal load
+	 * and actual load are not compatible for comparison. However, the
+	 * ratios of them are directly comparable. 
+	 */
+	prio = get_nice_prio(p);
+	if (prio >= NICE_WIDTH)
+		prio = NICE_WIDTH - 1;
+	else if (prio < 0)
+		prio = 0;
+
+	weight = sched_prio_to_slice_weight[prio];
+	return weight;
+}
+
+static u64 calc_task_load_actual(struct task_ctx *taskc)
+{
+	/*
+	 * The actual load is the CPU time consumed in a time interval, which
+	 * can be calculated from task's average run time and frequency.
+	 */
+	const s64 interval_adj = LAVD_TIME_ONE_SEC / LAVD_CPU_UTIL_INTERVAL_NS;
+	return (taskc->run_time_ns * taskc->run_freq) / interval_adj;
+}
+
+static u64 calc_slice_share(struct task_struct *p, struct task_ctx *taskc)
+{
+	/*
+	 * Task's CPU time share within a targeted latency window is basically
+	 * determined by its nice priority (and its corresponding weight). In
+	 * addition, if a task is compute-bound with high slice_boost_prio, the
+	 * scheduler tries to allocate a longer time slice.
+	 */
+	u64 share = get_task_load_ideal(p);
+
+	share += (share * taskc->slice_boost_prio) / LAVD_SLICE_BOOST_MAX;
+
+	return share;
+}
+
+static u64 calc_time_slice(struct task_struct *p, struct task_ctx *taskc)
+{
+	struct sys_cpu_util *cutil_cur = get_sys_cpu_util_cur();
+	u64 slice, share, gr_ft;
+
+	/*
+	 * The time slice should be short enough to schedule all runnable tasks
+	 * at least once within a targeted latency.
+	 */
+	share = calc_slice_share(p, taskc);
+	slice = (share * nr_cpus_onln) *
+		(LAVD_TARGETED_LATENCY_NS / cutil_cur->load_ideal);
+
+	/*
+	 * Take the task's greedy ratio into consideration. We assign a shorter
+	 * time slice when the task is greedy but not latency-critical.
+	 */
+	gr_ft = calc_greedy_factor(taskc);
+	slice = (slice * 1000) / gr_ft;
+
+	/*
+	 * Keep the slice in [LAVD_SLICE_MIN_NS, LAVD_SLICE_MAX_NS].
+	 */
+	if (slice < LAVD_SLICE_MIN_NS)
+		slice = LAVD_SLICE_MIN_NS;
+	else if (slice > LAVD_SLICE_MAX_NS)
+		slice = LAVD_SLICE_MAX_NS;
+
+	/*
+	 * If a task has yet to be scheduled (i.e., a freshly forked task or a
+	 * task just under sched_ext), don't give a fair amount of time slice
+	 * until knowing its properties. This helps to mitigate potential
+	 * system starvation caused by massively forking tasks (i.e., fork-bomb
+	 * attacks).
+	 */
+	if (!have_scheduled(taskc))
+		slice >>= 2;
+
+	taskc->slice_ns = slice;
+	return slice;
+}
+
+static bool transit_task_stat(struct task_ctx *taskc, int tgt_stat)
+{
+	/*
+	 * Update task loads only when the state transition is valid.
+	 *
+	 * initial state
+	 * -------------
+	 *     |
+	 *    \/
+	 * [STOPPING] --> [ENQ] --> [RUNNING]
+	 *    /\                        |
+	 *    |                         |
+	 *    +-------------------------+
+	 */
+	const static int valid_tgt_stat[] = {
+		LAVD_TASK_STAT_ENQ,
+		LAVD_TASK_STAT_RUNNING,
+		LAVD_TASK_STAT_STOPPING,
```

Comment:
Also, what were the transition cases which didn't follow the above transition diagram? I'm having a bit of hard time thinking up a case where it would deviate from the above.

Reply:
That's a good suggestion. That is much easier to read. I will change it as you suggested.

Two types of invalid state transitions have been observed so far, and there are reasons for that. The two are as follows:

- STOPPING -> RUNNING: This can happen when the task is forked (STOPPING) and then scheduled (RUNNING) without enqueuing.
- ENQ -> ENQ: The underlying scx core can issue enqueues more than once for a task.

Reply:
ENQ -> ENQ transitions are because `ops.dequeue()` is not implemented. What's actually happening is ENQ -> DEQ -> ENQ which happens when task attributes are updated (e.g. nice level, allowed cpus and so on). The DEQ cycling is always transient, so ignoring this is completely fine and will result in the correct behavior for lavd.

Reply:
The STOPPING -> RUNNING transitions are because there are several cases where scx core skips enqueue including:
- bypass mode is turned on (this is turned on during both init and exit. it's also used across suspend/resume operations.
- `SCX_OPS_ENQ_EXITING` is not set and an exiting task was woken up.
- The associated CPU is not fully online.

These events should be rare so the impact is limited but the current code is probably not behaving as intended right now as it'd end up ignoring STOPPING -> RUNNING transition completely when what's actually happening is more akin to STOPPING -> ENQ -> RUNNING where ENQ step is implicit.

I wonder whether  a better way to do this is using `ops.runnable()` and `ops.quiescent()` to add and subtract from system load numbers. They are guaranteed to follow `runnable ->  [ running -> stopping -> ]* quiescent` transitions and using them would make both ENQ -> ENQ and STOPPING -> RUNNING go away  and thus simplify the resulting code.

Reply:
Thanks for the detailed explanation! I updated the comments based on your explanation. 

For the STOPPING->RUNNING transition, actually I tried this time but it does not help improving the accuracy of task statistics. I thnink that's because they are rare and special cases.

Using ops.runnable() and ops.quiescent() is not the information that scx_lavd wants to collect. The goal is to collect task's actual CPU time not how long the task has been runnable. So I think accumulating time between ops.running() and ops.stopping() should be what we want.